src/interleave-generator.mjs

```diff
@@ -0,0 +1,42 @@
+import { ensureIterable } from './internal/iterable'
+import map from './map'
+
+class InterleaveBuffer {
+  constructor (iterator) {
+    this._iterator = iterator
+    this._next = this._iterator.next()
+  }
+
+  take () {
+    const next = this._next
+    if (next.done) {
+      return undefined
+    }
+
+    this._next = this._iterator.next()
+    return next.value
+  }
+
+  canTake () {
```

Comment:
**[DISCUSSION]** Should `canTake` be a property getter (or a property) instead of a function? Because tagged union would prove to be useful in TypeScript:

```typescript
// declaration

interface InterleaveBufferThatCanTake<T> {
  canTake: true
  take (): T
}

interface InterleaveBufferThatCannotTake {
  canTake: false
}

type InterleaveBuffer<T> = InterleaveBufferThatCanTake<T> | InterleaveBufferThatCannotTake
```

```typescript
// usage

if (buffer.canTake) {
  yield buffer.take() // no errors
} else {
  yield buffer.take() // TSC: Property 'take' does not exist on type 'InterleaveBufferThatCannotTake'
}
```

Reply:
On a second thought, this would probably not necessary, but TS declaration need to be tweaked:

```typescript
// declaration

interface InterleaveBuffer<T> {
  canTake (): this is InterleaveBufferThatCanTake<T>
}

interface InterleaveBufferThatCanTake<T> extends InterleaveBuffer<T> {
  take (): T
}
```

```typescript
// usage

if (buffer.canTake()) {
  yield buffer.take() // no errors
} else {
  yield buffer.take() // TSC: Property 'take' does not exist on type 'InterleaveBuffer<T>'
}
```

Reply:
I don't want to do that because I also decided not to make it an error to `take()` when `canTake()` is falsy. Basically if you don't check `canTake` only `canTakeAny` your interleave will end up filling missing values with `undefined`, which we do other places like `zipLongest` anyway. I believe that some people will want to do that. The tricky thing from the Typescript perspective though is that we don't capture the circumstance in which undefined may be introduced to the result type, so from that perspective I can see the attractiveness of using the types to force the user into a usage that the compiler can understand. I do think it's OK to have code that is more permissive in untyped usage than typed usage. Curious to hear what you think about this.

Reply:
Since the type of the final iterable does not contain `undefined`, I would like TypeScript to prevent me from fill the iterable with hidden `undefined` values.

Reply:
The thing is I don't want the error that the type system gives to be that the method `take` does not exist. That is actively misleading. Is there a way to express the idea that calling the function is a type error without hiding the whole function?

Reply:
But that would make it even more misleading IMO, so it is best to make it disappear.

Reply:
You can also make it returns `unknown` (3.2 feature) when `canTake()` return `true`.

Reply:
Okay, this is final: Return `T | undefined` when `canTake()` is `false`.

```typescript
// declaration

interface InterleaveBuffer<T> {
  canTake (): this is InterleaveBufferThatCanTake<T>
  take (): T | undefined // or null (I'm too lazy to double-check your code)
}

interface InterleaveBufferThatCanTake<T> extends InterleaveBuffer<T> {
  take (): T
}
```

This way, TSC will only screams error when `strictNullCheck` is set to `true`, just like the way TypeScript was designed.

Reply:
I just came to the same conclusion myself. https://www.typescriptlang.org/play/index.html#src=interface%20EmptyInterleaveBuffer%3CT%3E%20%7B%0D%0A%20%20%20%20canTake()%3A%20this%20is%20InterleaveBuffer%3CT%3E%0D%0A%20%20%20%20take%20()%3A%20undefined%20%7C%20T%0D%0A%7D%0D%0A%0D%0Ainterface%20InterleaveBuffer%3CT%3E%20extends%20EmptyInterleaveBuffer%3CT%3E%20%7B%0D%0A%20%20take%20()%3A%20T%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20buffer%3A%20EmptyInterleaveBuffer%3Cstring%3E%3B%0D%0A%0D%0Aconst%20taken%20%3D%20buffer.take()%3B%0D%0A%0D%0Aif%20(buffer.canTake())%20%7B%0D%0A%20%20%20%20const%20takenSafely%20%3D%20buffer.take()%0D%0A%7D

Reply:
I don't think "Empty" is necessary, a true `EmptyInterleaveBuffer` would have `take()` returns only `undefined`.

Reply:
Now the problem is that I still don't really want the return value of `canTake` to be a boolean for the reasons discussed previously. How bad do you think it would be if the type just misrepresents what is returned a bit?

Reply:
Also this approach just won't work for async.

Reply:
Then we should redesign `InterleaveBuffer`. Instead of having just one `class InterleaveBuffer`, we should have two:
  * One that provides `canTake` method that returns an object that provides `take`
  * One that provides `take`

And since both classes have just one method, we should return functions directly instead.

Reply:
If you leave Typescript out of the picture that is a worse API design than what we have. I don't want to create API designs that type systems will never be able to understand, but I also don't feel obligated to twist my APIs up because the current feature set of a type system doesn't give users the tools to verify that certain kinds of things are safe.

Reply:
I'm going to mark this resolved as I have updated this location with the type safety solution we discussed. We can still continue the discussion about how it should be handled for the async version, which I have left as it was.